Var ct

cryogenian opened the pull request:
Instead of pushing this thing to master it will be better to have different wip branch. 
This also makes it possible to make smaller prs! Yay, not 10k loc in one :)

garyb commented on src/SlamData/Workspace/Card/Ace/Component.purs:
```diff
@@ -32,6 +32,7 @@ import Control.Monad.Aff.AVar (makeVar, takeVar)
 import Data.Array as Array
 import Data.String as Str
 import Data.StrMap as SM
+import Data.Variant (case_)
```
May as well `case_` to the Prelude, since I guess we'll be using it quite a bit!

garyb commented on src/SlamData/Workspace/Card/CardType.purs:
```diff
@@ -15,310 +15,209 @@ limitations under the License.
+  , _cache, cache
+  , _draftboard, draftboard
+  , _geo, geo
+  , SimpleR, Simple)
```
It's probably better to keep these import as absurdly long single lines, since `psc-ide` / editors will reformat them that way anyway.

garyb commented on src/SlamData/Workspace/Card/CardType.purs:
```diff
@@ -15,310 +15,209 @@ limitations under the License.
+  , InputR, Input)
+import SlamData.Workspace.Card.CardType.Input as Inp
+
+import Utils (case2_)
```
If we add `case_` to the prelude, may as well add `case2_` too :smile:

garyb commented on src/SlamData/Workspace/Card/CardType.purs:
```diff
@@ -15,310 +15,209 @@ limitations under the License.
+        ( GeoR
+          ( SelectR
+            ( InputR
+              ( StaticR () ) ) ) ) ) ) )
```
I wonder if we can do this without needing the `R` types too, so we could union the variants directly... maybe would require the new `RowToList` stuff. Any thoughts @natefaubion?

(This isn't really a review comment, just speculation - we can always change it later if it is indeed possible).

garyb commented on src/SlamData/Workspace/Card/CardType/Ace.purs:
```diff
@@ -0,0 +1,88 @@
+  # on _aceSql (const "ace/mode/sql")
+
+eq_ ∷ ∀ r rr. (Variant r → Variant rr → Boolean) → Ace r → Ace rr → Boolean
+eq_ cb r = cb (unsafeCoerce r)
```
Let's make a typed function for whatever this `unsafeCoerce` is supposed to be doing - un-explicitly-typed-`unsafeCoerce` has always come back to bite me!

garyb commented on src/SlamData/Workspace/Card/CardType/Chart.purs:
```diff
@@ -0,0 +1,303 @@
+  "punch-card" → Right punchCard
+  "candlestick" → Right candlestick
+  "parallel" → Right parallel
+  _ → Left "this is not chart type"
```
How about:

```
ty → Left ("'" <> ty <> "'" is an unknown chart type")
```
for a marginally better error in the unlikely event it does occur?

garyb commented on src/SlamData/Workspace/Card/CardType/Geo.purs:
```diff
@@ -0,0 +1,89 @@
+geoHeatmap = inj _geoHeatmap unit
+
+eq_ ∷ ∀ r rr. (Variant r → Variant rr → Boolean) → Geo r → Geo rr → Boolean
+eq_ cb r = cb (unsafeCoerce r)
```
Same `unsafeCoerce` as before... in fact, I imagine every `eq_` is going to have one of these, so I'll skip making these comments any further :smile:

garyb commented on src/SlamData/Workspace/Card/CardType/Geo.purs:
```diff
@@ -0,0 +1,89 @@
+parse = case _ of
+  "marker" → Right geoMarker
+  "heatmap" → Right geoHeatmap
+  _ → Left "this is not geo chart card type"
```
Same suggestion for the `parse` error here in the catch-all case, and as with `eq_` I'll stop mentioning that from here down.

garyb commented on src/SlamData/Workspace/Card/Next/NextAction.purs:
```diff
@@ -38,7 +37,13 @@ data NextAction
+  eq (FindOutHowToInsert c1) (FindOutHowToInsert c2) = CT.eq_ c1 c2
+  eq _ _ = false
+
+--derive instance eqNextAction ∷ Eq NextAction
```
Commented out line

cryogenian commented:
@garyb At first I was thinking about making a pr with this and viztype picker and updated setup dsl, because variant isn't really necessary w/o viztype (in fact it's not necessary at all, but makes things a bit more convenient through reusing same funcs for different types)

cryogenian commented:
+ Moved `case_`, `case2_`, `expandVariant` to `SlamData.Prelude`. I don't think moving `on` is nice idea though. `on` from prelude is a bit more generic case. 
+ Added info to parse functions 
+ Made `eq_` returning `HeytingAlgebra`. In fact it's `Bounded` but there is no 
```purescript
instance boundedArr :: Bounded a => Bounded (r -> a) where 
  top _ = top
  bottom _ bottom
```
+ Also added names for `unsafeCoerce`s

garyb commented:
Changes look good, but I want to experiment a little with the composite `eq_` definitions, as they seem a little odd to me the way they are just now (in that sometimes you'd have to pass `case_` in or something like that, if you wanted to just compare `Ace ()`s).

I'll get that done this evening and comment back if I come up with something better.

garyb commented:
Actually, no this is best I think, without general `eq` for variants via `RowToList` (ahem @natefaubion)

If we define the subset-equalities with `T r -> T r -> Boolean` then the only way we can do that is by lying, in that it will return false for anything in `r`, even if it is equal. If we define them with `T () -> T () -> Boolean` then that will work, but in the super-`eq_` we need to `||` every sub-`eq_`, and also need to pre-match some stuff so it's more boilerplate and also more code to run when comparing.

Nicely done! :)

cryogenian commented:
@garyb ok, is this good to merge? :)

cryogenian commented:
BTW, I don't think `RowToList` will help for open variants, because it works only for closed rows

natefaubion commented:
If you try to use `RowToList` on open variants it will just defer the constraint, afaik.
